runtime/lua/vim/lsp/util.lua

```diff
@@ -1840,6 +1840,59 @@ function M.symbols_to_items(symbols, bufnr, position_encoding)
   return items
 end
 
+--- Converts links to quickfix list items.
+---
+---@param links lsp.DocumentLink[]
+---@param bufnr? integer buffer handle or 0 for current, defaults to current
+---@param position_encoding? 'utf-8'|'utf-16'|'utf-32'
+---                         default to first client of buffer
+---@return vim.quickfix.entry[] # See |setqflist()| for the format
+function M.links_to_items(links, bufnr, position_encoding)
+  bufnr = vim._resolve_bufnr(bufnr)
+  if position_encoding == nil then
+    vim.notify_once(
+      'links_to_items must be called with valid position encoding',
+      vim.log.levels.WARN
+    )
+    position_encoding = vim.lsp.get_clients({ bufnr = 0 })[1].offset_encoding
+  end
+
+  local items = {} --- @type vim.quickfix.entry[]
+  for _, link in ipairs(links) do
+    --- @type string?, lsp.Range?
+    local uri, range = link.target, link.range
+    if uri and range then
+      ---@type integer?, integer?, integer?, integer?
+      local lnum, col, end_lnum, end_col
+
+      ---@type string, string
+      local filename, text
+      if vim.startswith(uri, 'file:///') then
+        filename = vim.uri_to_fname(uri)
+        text = filename
+      else
+        filename = vim.uri_to_fname(vim.uri_from_bufnr(bufnr))
+        text = uri
+        lnum = range['start'].line + 1
```

Comment:
Nit:
```suggestion
        lnum = range.start.line + 1
```
We need to use `range['end']` because `end` is a Lua keyword.

Reply:
yes that is nicer